File: tries/208.py
class Trie:
    
    class TrieNode:
        def __init__(self):
            self.children = [None] * 26
            self.is_end_of_word = False
            
    def __init__(self):
        self.root = self.get_node()
        
    def get_node(self):
        return self.TrieNode()
    
    def insert(self, word: str) -> None:
        crawler = self.root
        
        for char in word:
            index = ord(char) - ord('a')
            if crawler.children[index] is None:
                crawler.children[index] = self.get_node()
            crawler = crawler.children[index]
        
        crawler.is_end_of_word = True
    
    def search(self, word: str) -> bool:
        crawler = self.root
        
        for char in word:
            index = ord(char) - ord('a')
            if crawler.children[index] is None:
                return False
            crawler = crawler.children[index]
        
        return crawler is not None and crawler.is_end_of_word
    
    def starts_with(self, prefix: str) -> bool:
        crawler = self.root
        
        for char in prefix:
            index = ord(char) - ord('a')
            if crawler.children[index] is None:
                return False
            crawler = crawler.children[index]
        
        return True

File: tries/208.js

class TrieNode {
  constructor() {
    // Array to store children nodes, one for each letter of the alphabet.
    this.children = new Array(26).fill(null); 
    // Boolean flag to mark if a word ends at this node.
    this.isEndOfWord = false; 
  }
}

class Trie {
  constructor() {
    this.root = this.getTrieNode();
  }

  getTrieNode() {
    return new TrieNode();
  }

  insert(word) {
    let crawler = this.root;
    for (const char of word) {
      const index = char.charCodeAt(0) - 'a'.charCodeAt(0);
      if (crawler.children[index] === null) {
        crawler.children[index] = this.getTrieNode();
      }
      crawler = crawler.children[index];
    }
    crawler.isEndOfWord = true;
  }

  search(word) {
    let crawler = this.root;
    for (const char of word) {
      const index = char.charCodeAt(0) - 'a'.charCodeAt(0);
      if (crawler.children[index] === null) {
        return false;
      }
      crawler = crawler.children[index];
    }
    return crawler !== null && crawler.isEndOfWord;
  }

  startsWith(prefix) {
    let crawler = this.root;
    for (const char of prefix) {
      const index = char.charCodeAt(0) - 'a'.charCodeAt(0);
      if (crawler.children[index] === null) {
        return false;
      }
      crawler = crawler.children[index];
    }
    return true;
  }
}